Userspace programs for extended Berkeley packet filter support

cthbleachbit opened the pull request:
Topic Description
-----------------

This PR introduces a handful of packages required to utilize eBPF tracepoints in kernel.

Package(s) Affected
-------------------

* `pahole`: DWARF debug symbol processor (by Lion Yang)
* `libbpf`: Library for eBPF loading from user-space
* `bcc`: eBPF program compiler collection (and python 3 bindings)
* `bpftrace`: High level BPF tracing utilities

Build Order
-----------

`pahole` | (`libbpf` -> `bcc` -> `bpftrace`) 

Architectural Progress
----------------------

- [x] AMD64 `amd64`   
- [x] AArch64 `arm64`

Secondary Architectural Progress
--------------------------------

- [x] Loongson 3 `loongson3`
- [x] PowerPC 64-bit (Little Endian) `ppc64el`

Post-Merge Architectural Progress
---------------------------------

- [x] AMD64 `amd64`   
- [x] AArch64 `arm64`

Post-Merge Secondary Architectural Progress
-------------------------------------------

- [x] Loongson 3 `loongson3`
- [x] PowerPC 64-bit (Little Endian) `ppc64el`

LionNatsu committed:
pahole: new, 1.19

cthbleachbit committed:
libbpf: new, 0.3

cthbleachbit committed:
bcc: new, 0.18.0

cthbleachbit committed:
bpftrace: new, 0.11.4

MingcongBai commented on extra-admin/bpftrace/autobuild/patches/0001-bpftrace-0.11.4-old-kernels.patch:
```diff
@@ -0,0 +1,28 @@
+diff --git a/src/btf.h b/src/btf.h
```
The naming of this patch is remarkably bad.

MingcongBai commented on extra-devel/bcc/autobuild/build:
```diff
@@ -0,0 +1,14 @@
+mkdir build
```
So why not change `SUBDIR` in spec? Also, use absolute paths here.

cthbleachbit replied:
Note that this file exists solely to workaround the issue where PKGVER isn't defined in defines. Will use absolute paths though.

MingcongBai commented on extra-devel/bcc/autobuild/defines:
```diff
@@ -0,0 +1,9 @@
+PKGSEC=devel
+PKGDEP="llvm elfutils libedit ethtool luajit python-3 netaddr libbpf"
+BUILDDEP="flex bison"
+PKGDES="BPF Compiler Collection"
```
Could do with a better description.

MingcongBai commented on extra-admin/bpftrace/autobuild/defines:
```diff
@@ -0,0 +1,12 @@
+PKGSEC=admin
+PKGDEP="elfutils zlib llvm bcc libbpf"
+BUILDDEP="linux+kernel"
+PKGDES="High-level eBPF tracing utility"
```
Could do with a better description.

MingcongBai commented on extra-admin/bpftrace/autobuild/defines:
```diff
@@ -0,0 +1,12 @@
+        -DLIBOPCODES_LIBRARIES=${SRCDIR}/libopcodes.so
+"""
+
+ABSTRIP=0
```
Why not?

cthbleachbit replied:
added some explaination.

MingcongBai commented on extra-devel/bcc/autobuild/defines:
```diff
@@ -0,0 +1,9 @@
+BUILDDEP="flex bison"
+PKGDES="BPF Compiler Collection"
+
+NOLTO=1
```
Why not?

MingcongBai commented on extra-kernel/libbpf/autobuild/build:
```diff
@@ -0,0 +1,6 @@
+pushd src
```
Use absolute path here... also please consider tweaking `SUBDIR`.

MingcongBai commented on extra-kernel/libbpf/autobuild/build:
```diff
@@ -0,0 +1,6 @@
+pushd src
+make
```
abinfo is missing throughout.

MingcongBai commented on extra-kernel/libbpf/autobuild/build:
```diff
@@ -0,0 +1,6 @@
+make
+make install DESTDIR="${PKGDIR}" LIBSUBDIR=lib install install_headers
+popd
+
```
Verbosity please.

MingcongBai commented on extra-kernel/libbpf/autobuild/defines:
```diff
@@ -0,0 +1,7 @@
+PKGNAME=libbpf
+PKGSEC=kernel
+PKGDEP="elfutils linux+api"
```
Why is `linux+api` a dependency?

cthbleachbit replied:
My bad, should be builddep.

MingcongBai commented on extra-kernel/libbpf/autobuild/defines:
```diff
@@ -0,0 +1,7 @@
+BUILLDEP="rsync"
+PKGDES="Library for eBPF loading from user-space"
+
+NOLTO=1
```
Why not?

cthbleachbit replied:
`libbpf` does not build with LTO.

MingcongBai commented on extra-kernel/libbpf/autobuild/defines:
```diff
@@ -0,0 +1,7 @@
+PKGSEC=kernel
+PKGDEP="elfutils linux+api"
+BUILLDEP="rsync"
+PKGDES="Library for eBPF loading from user-space"
```
I'd suggest the following...

```
Library for loading eBPF functionalities/features/layers/... (?) from userspace
```

MingcongBai commented on extra-utils/pahole/autobuild/defines:
```diff
@@ -0,0 +1,6 @@
+PKGNAME=pahole
+PKGSEC=utils
+PKGDEP="elfutils zlib"
+PKGDES="Pahole and other DWARF utils"
```
Please do not use any shorthand in `PKGDES`.

MingcongBai commented on extra-devel/bcc/autobuild/defines:
```diff
@@ -0,0 +1,8 @@
+PKGDES="BPF compiler collection and tools for Linux kernel tracing and instrumentation"
+
+NOSTATIC=0
+
```
Trailing blank line.

MingcongBai commented on extra-kernel/libbpf/autobuild/defines:
```diff
@@ -0,0 +1,10 @@
+PKGSEC=kernel
+PKGDEP="elfutils"
+BUILDDEP="linux+api"
+BUILLDEP="rsync"
```
Typo, and you were probably on the way to override your last `BUILDDEP` declaration.

cthbleachbit committed:
bpftrace: backport upstream patch, add gdb to builddep

arm64 does not define SYS_open as most programs have moved to a more
modern openat(). This prevents one of the test program from building as
it directly interfaces with SYS_open. This has been fixed in master
branch, backported here as patch 0002.

bpftrace also requires /usr/include/diagnostics.h to exist at build
time, thus gdb in BUILDDEP.
